clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -478,14 +488,38 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
     } else if (type->isVectorType()) {
       llvm_unreachable("no vector inc/dec yet");
     } else if (type->isRealFloatingType()) {
-      auto isFloatOrDouble = type->isSpecificBuiltinType(BuiltinType::Float) ||
-                             type->isSpecificBuiltinType(BuiltinType::Double);
-      assert(isFloatOrDouble && "Non-float/double NYI");
-
-      // Create the inc/dec operation.
-      auto kind =
-          (isInc ? mlir::cir::UnaryOpKind::Inc : mlir::cir::UnaryOpKind::Dec);
-      value = buildUnaryOp(E, kind, input);
+      if (type->isHalfType())
+        llvm_unreachable("__fp16 type NYI");
+
+      if (value.getType().isa<mlir::cir::SingleType, mlir::cir::DoubleType>()) {
+        // Create the inc/dec operation.
+        auto kind =
+            (isInc ? mlir::cir::UnaryOpKind::Inc : mlir::cir::UnaryOpKind::Dec);
+        value = buildUnaryOp(E, kind, input);
```

Comment:
I'm a bit curious about the assumption of using unary ops here. I forgot about this and had to look above for `// NOTE(CIR): clang calls CreateAdd but folds this to a unary op`. All it takes for LLVM to fold `fadd` to unary is the value being single or double? Seems like an early optimization. Since this is the status quo for other handling that it's not specific to this patch, no worries to change it right now, but I'll like a comment here explaining why this is using a unary op and how this assumption works.

Reply:
Updated.